### DISCUSSION TITLE
Halbach array and multipole feature

### DISCUSSION THREAD

**Discussion by BillyTheKidPhysicist** [magpylib/magpylib#491] [Ideas]
Halbach arrays are configurations of permanent magnets that allow for field amplification/attenuation, and the construction of multipole magnets [1]. With Halbach arrays a magnetic field can be generated with off the shelf permanent magnets with values nearly twice the remnant field value. Halbach arrays also allow for the construction of low cost permanent multipole fields for use in accelerators or atom optics. A version of this would be great in magpylib, and I imagine many people would appreciate it. I have a simple version of a sextuple halbach array magnet composed of magpylib cuboids (previously boxes) that I actively use in my simulations.  It's a simple matter to implement and requires only rotating and positioning cuboids. 

Currently the field is calculated from each cuboid, but I think a more efficient approach would be to use a single cuboid and pass an array of coordinates which exploit symmetry. Additionally, a feature which searches for wether the remnance is being exceeded by fields would be a nice feature eventually because that is a problem with multiple layers. I've attached an image of a sextuple field in figure for an upcoming publication. 

Does this seem reasonable? 

1. Design of permanent multipole magnets with oriented rare earth cobalt material, Halbach, 198
![lensFieldSimulation](https://user-images.githubusercontent.com/55501990/157706150-8c5ad431-5ec5-42e3-9781-ef6b1eea1769.png)
0

**Comment by OrtnerMichael** [comment]
hi @BillyTheKidPhysicist 

Its my understanding that Discrete Halbach Arrays are mostly used to create strong homogeneous fields with off-the-shelf magnets. There are tons of papers with various designs that look like this

![image](https://user-images.githubusercontent.com/47384012/157980445-ef95fc05-2d63-430a-9393-f8761bfe769c.png)

Small and large diameters, a variety of magnet numbers (6-50) per ring, single rings, nested rings, stacks, and so on. But all Halbachs iv'e seen so far follow the original Halbach magnetization pattern Cos(2\phi) - and depending on how the (axially magnetized) magnets must be rotated, lower and higher filling fractions, fields and homogeneity are achieved.

Some Magpylib developers(myself, @Alexboiboi, @pribossek) have been using such arrays and we all have our own codes. I think we all agree that developing a Magpylib Halbach Class would be extremely beneficial.

Currently we are quite busy to complete the v4 release, but for v4.1 or 4.2 I think that there should be such a Halbach class - also some of our research goes in this direction.

However, this will require a lot of work in terms of
1. class API (reasonable properties, methods layout, subclasses, (e.g. Halbach Ring),...)
2. class implementation (how to integrate into Magpylib, (e.g. tools subpackage #413 ), inheriting Collections properly, custom 3D model optional display, v4 offers an extremely powerful compound philisophy when working with Collections that should function as basis (documentation will be uploaded in the next days))
3. computational efficiency (probably we don't need to worry about that if we just inherit the Collection class, this will not be the absolute fasted way to compute the field, but it enables treating Halbachs as compound objects (i.e. full Magpylib object functionality) which will have so many advantages in terms of user-friendliness

I think it would make a lot of sense to outline in this discussion what should be the format of the Halbach class, and I'd be happy if everyone participated in this discussion.

@BillyTheKidPhysicist 
what surprises me about your design is that the field amplitude is very small inside of some magnets of the inner ring. Also I would expect a much more homogeneous field inside the array. Can you show us a figure with the magnetization pattern ? (with the v4 plotly backend you can create paper-quality 3D graphics)

sincerely,
michael

**Reply by BillyTheKidPhysicist** [reply]
Thank you for the reply. Unfortunately I have not updated  Magpylib to version 4.0 yet, so I cannot easily make the magnetization plot you are requesting, though I have before for sanity checks. The image I attached is of a hexapole magnet, whereas the images you shared are of dipoles, so I do not expect a uniform field. I am intentionally constructing a non uniform field because I require a  harmonic potential, ie a potential that depends on r^2, where r is the distance from the origin in the xy plane (The plane that you see in the image of my magnet). This is equivalent to a lens under certain conditions for particles with a magnetic moment. We are currently using such a lens to focus a beam of atoms. The lens works as expected experimentally. I'll attach a neat photo of it. You can see the layers levitating as we are beginning the perilous job of assembling. 
[IMG_1666.pdf](https://github.com/magpylib/magpylib/files/8236590/IMG_1666.pdf)

**Reply by OrtnerMichael** [reply]
Hi @BillyTheKidPhysicist 

This is very interesting, thank your for sharing ! - I must confess I was not aware of this application.

What you propose above (making use of the symmetry) will work for sure. For maximal performance you can use direct core access with version 4 (btw it installs with `pip install magpylib --pre`). The steps would be as follows:

1. You need a rotation invariant grid
- Rotation invariant means that grid points will conincide for all magnets when you rotate afterwards
- You only need half of the complete grid because of the mirror symmetry, 
- Bring the grid to shape (n,3).
2.  With core access there is no relative position and orientation anymore (the magnet is in the center) - so you need a grid offset in one direction.
3. Compute the field on the grid using `magpy.core.magnet_cuboid_field(x x x x x)`. Details on arguments are in the docstrings.
4. Mirror the grid using only numpy commands
5. Copy/rotate the field for every magnet instance (the `scipy.spatial.transform.Rotation` class is great btw)
6. Finally superpose

Some questions:
1. Why do you need so much performance ? Just using the magpylib object oriented interface will probably be a bit slower but its so much more convenient to use, and allows you to include fabrication tolerances with almost no effort.
2. Do you account for fabrication tolerances ? Do you use shimming ? 
3. Would it be ok to write a simple example code together that we can publish in the magpylib doku example galleries ?

Also:
I still find it odd that you have these small magnets with no field inside them :). If this is the case it means that you have an extremely strong opposing field there and you should account for material interactions.

**Reply by BillyTheKidPhysicist** [reply]
1: Currently, I am designing a storage ring (>7 meter round trip length) for neutral atoms composed of permanent magnets. We have a novel source of neutral atoms we have developed, the subject of an upcoming paper citing Magpylib, which is used to load this storage ring. The problem at had is to find a satisfying configuration in parameter space that allows for vacuum limited (10-100s) storage times. It is largely black box optimization , and the two bottlenecks are generating fields with Magpylib and compiling Numba functions and classes, so performance here would be highly relevant.

2a: I am in the progress of coding up a system to account for realistic misalignment error to ensure the storage ring is stable once constructed.

2b: This is a whole other issue that I am working on right, I hope to have a paper about it soon! It turns out that hexapole atom lenses have strong spherical aberration. This is not present when charged particles interact with multipole magnetic fields (here the quadrupole is the lens ). Halbach shows this in his original paper I believe. Basically, the field value averaged length (integral of Bx, By, or Bz through the magnet divided by the peak value of Bx, By, Bz respectively ) of a magnetic pole (dipole, quadrupole, etc) is constant for any radius. If however you are interested in the gradient of the norm, it has a strong dependence as a function of radius from origin. This results in a spherical aberration. It is very annoying for us. I am working on using shims to reduce this, and have had some luck (~90% reduction in spot size)

3: Yes! I already have a wrapper class that allows me to generate hexapole lens fields from cubic Magpylib magnets. I think what you might want though is a general approach that when given a n order,  constructs the configuration of magnets that generate that pole. I've never collaborated like this (except with my immediate colleagues). What should I do?

Yes, it is weird! But we accounted for that by using higher coercivity magnets on the interior based on the opposing fields. Is it possible we are missing something here? The idea was to have the BH curve still be "flat". It must work well enough cause I can predict where our particles end up quite accurately .I've attached a photo of the magnetization, and field simulation from Comsol we used earlier on (before abandoning Comsol). You can see from the magnetization why every other inner magnet has a strongly reduced field. For fun as well, here is a camera photo of our atoms being focused by the lens and fluorescing. (It was probably about midnight after 12 hours of data taking
<img width="993" alt="focus" src="https://user-images.githubusercontent.com/55501990/158066846-9cbd69db-e801-4684-aa92-9b1dc8f30cf6.png">
 XD )
<img width="745" alt="Screen Shot 2022-03-13 at 10 15 35 AM" src="https://user-images.githubusercontent.com/55501990/158066541-274ccd23-0610-453e-9676-332b84d580ea.png">
<img width="803" alt="Screen Shot 2022-03-13 at 10 15 40 AM" src="https://user-images.githubusercontent.com/55501990/158066544-b045330f-f018-4e73-90f5-472da7b3434e.png">

**Reply by BillyTheKidPhysicist** [reply]
I think you are right about the performance. Looping through cuboids and computing the field isn't that different than some wizardy with coordinates and a single cuboid

**Reply by OrtnerMichael** [reply]
hi @BillyTheKidPhysicist 

1. cool stuff :) ... im a theorist who never went to the lab, so I'm really impressed by this 👍 

2. In terms of material interaction: Even when you have high coercivity magnets, and never sample the non-linearity, you still have to deal with the linear part of the demagnetization curve, often referred to as remanent permeability, or just permeability. Its a simple linear, reversible response to an opposing field. For high grade materials the curve is flat, \xi<0.05, but it should not be neglected. When doing cuboid analytics, one can estimate this influence by determining the mean H-field inside the cuboid, and adjusting the magnetization according to the material law. This is called "demagnetization factor" and there is an analyical expression for cuboids. With multiple cuboids you will need to solve a matrix equation, because the cuboids influence each other. If you are interested in this I propose you read up on "magnetostatic method of moments". I also have a code you could use for this purpose - let me know if you are interested.

3. So when you construct your lenses I imagine you will have a lot tolerances - the biggest influence is probably the magnet magnetization. We have seen 5-15 % variation there, up to 3 deg angle error from industrial sources. The performance code is probably quite a lot faster(mabye factor of 2 or 3), but the general code can account for all fabrication tolerances, and code development time is minimal. And why would you develop a code that will only describe perfect lenses ? Are you using the object oriented or direct interface (getBv in version 2 and getBdict in 3) ? With the direct one you can get some extra performance.

4. As you mentioned black-box optimizer ... we have made some very good experiences with differential evolution optimizer, up to 30 variables on our 48 core intel skylake server, with 1-3 hours computation time only - so I guess there is a lot of room to go higher. What are you using ?

5. Concerning the collaboration idea: In our new documentation (will be ready any day now) we have an example section. I would create a new example e.g. "Magnetic hexapole lens design". And we could design a very simple code, have some nice graphical representation, some descriptive text, brief explanations (how does a magnetic lens work, what is it used for...), maybe a foto of your experiment, a link to your group or your paper if you want, all on the magpylib doku webpage. The format is sphinx 4 flavored markdown ... so basically markdown with some additional features like code compilation. I would support you with the code and markdown page development ... I can help out with graphics (like this one but interactive). The idea is just to show some cool and useful application and code.

![image](https://user-images.githubusercontent.com/47384012/158694940-015515a9-390c-439a-9a14-83111f2e5894.png)

**Reply by BillyTheKidPhysicist** [reply]
1: I'm glad you like it! Honestly I feel like I'm in over my head most of the time 😅

2: Yes, I am absolutely interested, thank you, and thank you for bringing this up, I have been neglecting this contribution. I hope it doesn't through a wrench in everything... We have experimentally verified our existing lens already and it appear to work mostly as expected, but there is some discrepancy. We are chalking that up to not understand the dynamics of our source well enough (a seed edsupersonic ~millikelvin jet)

3: I'm sorry, I sort of lost you here. Are saying that Magpylib can include these effects of imperfect magnets? I definitely think developing a realistic code that would model errors would be very helpful@ I am using magpylib.magnet.getB in my wrapper class. 

4: Sort of embarrassing, but I am using my MacBook with the new m1 chip 😅. It actually works pretty well. I also have access to a 16 core AMD cpu. My optimization problems take a few days so it's not too bad. I looked into using the cloud or the university supercomputer, but so far I have plenty to work on in parallel as they run and it has not been a bottle neck. I'm a big fan of DE as well and I wrote an asynchronous differential evolution solver that works pretty well. I tried using gradient descent with momentum and Adam but the landscape is too jagged and lumpy. Gaussian Process wasn't that helpful either, and a little opaque to me, so I just stuck with DE. May I ask what you are optimizing out of curiosity? My shimming optimization get's up into the high 10's of dimensionality. Sorry, don't mean to ramble!

5: I would love to help if I can be of any! It might helpful to some people at some point to include particle tracing and have a sort of atom optics library, but that's probably pretty ambitious. 

Thank you so much for taking the time to talk to me.

**Comment by OrtnerMichael** [comment]
**2+3**: Its likely that the symmetry of your problem results in a similar lens field amplitude extremum position when you include material interaction, but that the field gradient ("lens curvature" that you also show above) is changed a little. Yes you can deal with this problem with Magpylib at 3 stages - each stage leads to more precision, but also more computational effort.

**First stage**: only include the magnet self-interaction by simply reducing the magnetization by the [demagnetization factor](http://www.magpar.net/static/magpar/doc/html/demagcalc.html)

**2nd stage**: Include the magnet-magnet interaction. You have 24 magnets, this will result in 24x24 matrix diagonalization (piece of cake even for your m1). This will fix the homogeneous contribution to the demagnetization which typically accounts for a large part of the effect.

**3rd stage**: Split up every Magnet into a lot of small pieces and compute the compute the full interaction. If you split up every magnet into 10x10x10 pieces you will have to diagonalize a dense 24000x24000 matrix ... which is where things become slow. I solve 10k pieces (in total) within a few minutes on my 4 core i5 laptop.

![image](https://user-images.githubusercontent.com/47384012/158890987-08432f30-4040-498a-ab18-a991cfc62f50.png)

In the worst case you can let the solver run for a while for some fine discretization, just to understand your error. I'll review the code that I have tomorrow and send it to you by email - it's kind of simple - you just apply it to a Collection, hand it a material parameter xi, and it adjusts the object magnetization vectors.

**4** : We optimize and calibrate magnetic position systems. We always have magnets and sensors around with a lot of tolerances - usually not more than 20-30. For this DE works like a charm. [pygmo](https://esa.github.io/pygmo2/) is a pretty cool package if you want more sophisticated DE and more parallelization.

**5**: Well I imagine just a short story to get ppl interested. And its always cool to know that this is real stuff - so a nice photo would help. An outline could be

**Magnetic lenses for particle accelerator and storage rings**
- 5 lines on particle storage rings
- 5 lines on what is a magnetic lens and how does it work, maybe a photo of a real lens
- A simple Magpylib code that shows the field distribution and the geometry
- Maybe an outline of the lenses of a storage ring ? with atomic potential, like i tried to draw it upstairs :)
- Maybe the photo you included above where the atoms actually become visible
- finally maybe a link to your publication, and/or your group

We could make a magpylib github branch and just add a blank example file to the doku where you can work. If you feel uncomfortable with markdown you could just post the text in here and I add it, or send me a text file. Probably I would help with magnetic field visualization - maybe we do something like this:

![image](https://user-images.githubusercontent.com/47384012/158893941-60225746-c92c-452c-9475-0fdd1c2d66b0.png)

**Reply by BillyTheKidPhysicist** [reply]
wow, thank you very much, that would be great! And your description of the stages of improvement are very helpful.

I'm mild comfortable with markdown, but I should improve so I have no problem making it myself. I'll work on assembling a story as you have described. Our paper is not out the door yet, hopefully early next month so that part may have to wait. Working on a new branch would be great.

Thank you so much again for your help

**Reply by OrtnerMichael** [reply]
Well ... i messed something up from my old code - the results look good at first glance, but comparison to FE shows that something is off. I modified the solver - this will take a few days to fix, sorry.

**Reply by BillyTheKidPhysicist** [reply]
Not a problem at all!

**Reply by BillyTheKidPhysicist** [reply]
I have a quick question if you don't mind. For the 2nd stage you described above, including the magnet-magnet interaction, would it be a reasonably valid approximation to instead of solving an NxN matrix, consider it to be decoupled? (sort of a step 1.5) By this I mean say I have N magnets, for each magnet I consider how the N other magnets (including itself) affect its magnetization, but then neglect how this change affects the rest of the N-1 magnets. One could imagine sweeping through the N magnets with this procedure several times or some other technique like applying updates only after addressing each magnet. Perhaps I am simply describing a poor FEM solver

**Comment by BillyTheKidPhysicist** [comment]
In terms of fitting into Magpylib , should I think of a Halbach array (in this case a hexapole lens to get started) as a wrapper around a Collection object, or some other abstraction? I'm refactoring my code to apply the method of moments to a bending segment. My previous approach to the bending segment did not allow for this in a reasonable manner.

**Reply by BillyTheKidPhysicist** [reply]
Actually, I mean inheriting from Collection

**Reply by OrtnerMichael** [reply]
hi @BillyTheKidPhysicist 

We are currently reworking the Collection class, and preparing a sort of guideline how to properly subclass it with custom parameters and dynamic 3D models. This will be released in the next few days/weeks. I propose that you wait until then before you start to implement something.

Meanwhile it would be helpful to outline what the class would look like - i.e. to define the API, which will require a lot of discussion in terms of which parameters should be there and how should they be implemented. E.g. there could be a central Halbach object class to which Halbach rings can be added by methods

```python

halbach = Halbach()

halbach.add_ring(
   n=15,
   radius=10,
   cube_dimension=(1,1,1),
   magnetization = ?,
   z_position=0,
   ...
)
```

**Reply by BillyTheKidPhysicist** [reply]
Got it! 

At atom lens is a n=2 multipole,  an atom prism (where speed or high/vs low field seeking is the analogous to wavelength) is a combination of n=0 and n=2.  Perhaps a class which can implement any combination of multipoles by constructing the appropriate arrangement of permanent magnets. Cuboids as first, but maybe eventually prisms as well?

Then a class which can use the above class to build common helpful optics, such as atom lens, atom prism, atom waveguide (a combination of atom lens). I've attached an image of a bending atom waveguide
<img width="569" alt="Screen Shot 2022-03-26 at 9 09 03 AM" src="https://user-images.githubusercontent.com/55501990/160243241-5bf4b344-06d1-4381-9277-1eb6b7683dcb.png">

**Reply by OrtnerMichael** [reply]
ok so what properties would one ring have ?

- magnet base form, str, e.g. 'Cuboid', 'Cylinder', ...
- number of magnets, int n, (equidistant spacing by default)
- magnet dimension, array_like, shape (d,) or (n,d), single dimension or array_like with n individual dimensions to deal with variations between magnets
- ring radius, float
- ring position on z-axis, float
- halbach magnetzation type: int? (2pi, 4pi, 6pi ...), i think this would cover our use cases
- individual magnet position variations, array_like, shape (n,3), should probably be in cylindrical coordinates (dr, dphi, dz)
- individual magnet orientation variations, scipy rotation object with length n

i think this would cover it for our purposes ... what do you think ?

**Reply by BillyTheKidPhysicist** [reply]
Sounds good! Though it appears @pribossek  beat me to the punch 😅. Another helpful feature, and essential for some users, would be a superposition of multipoles. 

Halbach magnetization type is multipole order?

**Comment by pribossek** [comment]
Hello all,

I have the complete Halbach class implementation covering more or less  the
entire functionality summarized in the last post from Michael. Few extra
goodies are included as well. The class was written to work with a
pre-release of the MagPylib3, so probably it has to be slightly adapted to
newest changes introduced with the latest version, which I have not been
closely following over the last year.
@Michael: if there is enough interest, I can file a PR or we can review the
functionality in the upcoming days together.

Regards,
Jaka Pribosek










V V sob., 26. mar. 2022 ob 19:10 je oseba mortner ***@***.***>
napisala:

> ok so what properties would one ring have ?
>
>    - magnet base form, str, e.g. 'Cuboid', 'Cylinder', ...
>    - number of magnets, int n, (equidistant spacing by default)
>    - magnet dimension, array_like, shape (d,) or (n,d), single dimension
>    or array_like with n individual dimensions to deal with variations between
>    magnets
>    - ring radius, float
>    - ring position on z-axis, float
>    - halbach magnetzation type: int? (2pi, 4pi, 6pi ...), i think this
>    would cover our use cases
>    - individual magnet position variations, array_like, shape (n,3),
>    should probably be in cylindrical coordinates (dr, dphi, dz)
>    - individual magnet orientation variations, scipy rotation object with
>    length n
>
> i think this would cover it for our purposes ... what do you think ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/magpylib/magpylib/discussions/491#discussioncomment-2443802>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIHNA26W5QXNFOZFNYFQS3VB5HINANCNFSM5QNDGDGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


--

**Comment by OrtnerMichael** [comment]
Hi @pribossek and @BillyTheKidPhysicist 

I think it would be great to work on such a PR together. @pribossek my proposal is mostly based on the implementation you showed me some time ago. But I think we should first discuss all the features in detail - there might be some missing that would be critical for the problem of @BillyTheKidPhysicist, who introduces yet another new use case.

Also, the new Halbach class should follow the compound philosophy outlined in the new Magpylib guidelines. They are not published yet - but we are very close to the release. Once its out I would like to focus more on the Halbach problem.

@pribossek can you outline what functionality is missing in my first proposal ?

@BillyTheKidPhysicist  can you please review the functionality and comment on missing features that come to your mind ? What do you mean by "superposition of multipoles" ?

**Comment by pribossek** [comment]
@mortner I can confirm that current class implementation include all
functionality you have summarized. There are some extra goodies, in-built
export to CAD, in-built methods for magnetic field evaluation
(inhomogeneity and some other statistics) etc. Some simple and fast
in-built geometry optimizer might be nice too. This was already
implemented, but later removed because of increased class complexity.
Still, I believe it can be useful in daily work with Halbach systems.

In my view, what is missing is the following:
Current class constructs systems in the XY plane. I can imagine, a real
application would demand to arrange several such Halbach rings into a 3D
arrangement as in @BillyTheKidPhysicist example few posts above. This is
currently not supported in my class.
What a user can do of course is to use Halbach class to generate a
Collection and further manipulate this collection using in-built scipy
rotation class. However, most of the functionality of the Halbach class is
then lost (field evaluation, export, individual magnet pose definition..)
Yet, I think it might be a benefit, if we expand current Collections to be
able to take Halbach objects as well. This should probably be quite easy,
because Halbach objects are Collections themselves. I am not sure if in
current MagPyLib version Collections class can manipulate subcollections,
but I imagine this would find also many other nice applications. This way,
one would have the possibility to arrange group of rings in a 3D space
freely, yet maintain the functionality to manipulate each such ring in its
respective native plane.

@mortner: I don't know what you mean by compound philosophy. Can you
elaborate a bit?

@BillyTheKidPhysicist: yes, multipoles are supported.

V V tor., 29. mar. 2022 ob 22:27 je oseba mortner ***@***.***>
napisala:

> Hi @pribossek <https://github.com/pribossek> and @BillyTheKidPhysicist
> <https://github.com/BillyTheKidPhysicist>
>
> I think it would be great to work on such a PR together. @pribossek
> <https://github.com/pribossek> my proposal is mostly based on the
> implementation you showed me some time ago. But I think we should first
> discuss all the features in detail - there might be some missing that would
> be critical for the problem of @BillyTheKidPhysicist
> <https://github.com/BillyTheKidPhysicist>, who introduces yet another new
> use case.
>
> Also, the new Halbach class should follow the compound philosophy outlined
> in the new Magpylib guidelines. They are not published yet - but we are
> very close to the release. Once its out I would like to focus more on the
> Halbach problem.
>
> @pribossek <https://github.com/pribossek> can you outline what
> functionality is missing in my first proposal ?
>
> @BillyTheKidPhysicist <https://github.com/BillyTheKidPhysicist> can you
> please review the functionality and comment on missing features that come
> to your mind ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/magpylib/magpylib/discussions/491#discussioncomment-2464821>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIHNA6L7OS5I67OJSNSDOLVCNRRFANCNFSM5QNDGDGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


--

**Reply by OrtnerMichael** [reply]
@pribossek 

the "compound philosophy" is pretty much what you have expressed:

In the new magpylib 4 version:

- Collections are objects like any other object. They span a local reference frame for their children and have position and orientation attributes. This means that you can move Collections around like other objects and track their position. With these features we can do exactly what @BillyTheKidPhysicist needs to construct a set of individual rings with arbitrary positions and orientations.

- Collections can now contain other collections. This means that `Halbach` could be a super class, like in your original code, that contains multiple `Ring` objects, each of which is a Collection.

- A compound is class that sub-classes Collection. In that sense `Ring` and `Halbach` should both be compounds. This makes it now possible to treat a `Ring` or `Halbach` instance like a native Magpylib object, with `getBH`, `move` and `rotate` and even custom 3D graphical representation.

It would be great if we could outline and describe all desired properties and methods of `Ring` and `Halbach`. Maybe you can post your code here @pribossek ?

**Comment by BillyTheKidPhysicist** [comment]
@OrtnerMichael 
By superposition of multipoles I mean a halbach ring which is not a pure dipole or quadrupole or hexapole etc, but rather has components of each multipole. Below is an image of a realization of a dipole+quadrupole magnet, often referred to as a "combined function" magnet. In charged particle accelerators these have been used to reduce the length of lattices by combing two features into a single element: bending + focusing. This is typically done with ferromagnets or superconductors, permanent magnets are uncommon but interest is growing for low energy accelerators. A dipole+quadrupole halbach ring functions as a very good Stern-Gerlacht magnet for neutral atoms. 

@pribossek
Your implementation covers combination of multipoles, like in the image below? 

source: [Halbach Magnets for CBETA, D. Trbojevic et al., Brookhaven National Lab]

![Uploading combinedFunction.png…]()

**Reply by OrtnerMichael** [reply]
I cannot open that image, but I went through that reference. When you say dipole, quadrupole, hexapole etc. you refer to the field in the center of a ring, where your atoms are flying through, right ?

In the original Halbach paper a hollow Cylinder has a magnetization that depends on the azimuth θ described by 

Mr = M_0 cos((k − 1)θ) sin((k − 1)θ)
Mz = M_0 sin((k − 1)θ) sin((k − 1)θ)

where k describes the "poles".

I'm putting multipoles on the above list. This should not be difficult.

**Reply by BillyTheKidPhysicist** [reply]
I don't know what I did wrong. I did the same thing I did with the other images. Here's a dropbox transfer link: https://www.dropbox.com/t/qhzMbaLVf0tScTa1

I will try attaching the image again'

yes, dipole, quadrupole etc are interior to the ring where the atom fly through in my case. A perfect arrangement of Mr and Mz for a given k in an infinitely long cylinder would have a single multipole coefficient everywhere inside. Of course in real life there will be other multipole terms. Here's a great recent paper if you are interested: https://www.dropbox.com/t/KP3TtyMTvHYjZbcb
 




![combinedFunction](https://user-images.githubusercontent.com/55501990/161271308-7f38f5ed-9af9-47d8-a323-904f6be65c75.png)

**Comment by OrtnerMichael** [comment]
New list (because the other one is hidden in the answers)

- magnet base form, str, e.g. 'Cuboid', 'Cylinder', ...
- number of magnets, int n, (equidistant spacing by default)
- magnet dimension, array_like, shape (d,) or (n,d), single dimension or array_like with n individual dimensions to deal with variations between magnets
- pole-number, int k, Halbach pole number
- ring radius, float
- ring position on z-axis, float
- individual magnet position variations, array_like, shape (n,3), should probably be in cylindrical coordinates (dr, dphi, dz)
- individual magnet orientation variations, scipy rotation object with length n

**Reply by BillyTheKidPhysicist** [reply]
I just finished my thesis! I see that there is discussion of a new "tools" feature. Is anyone working on a Halbach class?

**Reply by OrtnerMichael** [reply]
hi @BillyTheKidPhysicist 

Sorry of the late reply and congratulations on your thesis. At the moment progress is slow here - we are mostly working on the Facet class to realize arbitrary geometries. The content of the "tools" package was not discussed yet. Maybe it can be a separate package at this point that can be integrated into magpylib in the future. But feel free to open an issue and start a discussion 👍 

Would you still write and application example for the documentation ?

mfg Michael
